### PR TITLE
[query] Add /namespace/ready endpoint

### PR DESCRIPTION
**Pull request by nbroyles**
**What this PR does / why we need it**:

Endpoint is to be used to signal that namespace is ready to receive
reads and writes from the coordinator.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/query/api/v1/handler/namespace/mark_ready.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	xhttp "github.com/m3db/m3/src/x/net/http"
+	"github.com/m3db/m3/src/x/retry"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+var (
+	// M3DBMarkReadyURL is the url for the M3DB namespace add handler.
+	M3DBMarkReadyURL = path.Join(handler.RoutePrefixV1, M3DBServiceNamespacePathName, "mark_ready")
```
Comment (ryanhall07) [inline src/query/api/v1/handler/namespace/mark_ready.go:51]: typically you wouldn't include a verb like `mark` in the path and instead rely on POST to mean you updating the state. so `POST` -> /ready would change it. and `GET` -> /ready would read it.

File: src/query/api/v1/handler/namespace/mark_ready.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+	// MarkReadyHTTPMethod is the HTTP method used with this resource.
+	MarkReadyHTTPMethod = http.MethodPost
+
+	fetchRetryConfig = retry.Configuration{
```
Comment (ryanhall07) [inline src/query/api/v1/handler/namespace/mark_ready.go:56]: seems like it would be better to push this retry to the client who is calling the endpoint

File: src/query/generated/proto/admin/namespace.proto [hunk 3]
```diff
@@ -49,3 +49,19 @@ message NamespaceSchemaResetRequest {
...
 }
+
+// NamespaceMarkReadyRequest is the request for transitioning a namespace
+// to the ready state.
+message NamespaceMarkReadyRequest {
+  // Name is the namespace name.
+  string name = 1;
+  // Force forces the StagingStatus of the namespace
+  // to be moved to the ready state.
+  bool force  = 2;
```
Comment (ryanhall07) [inline src/query/generated/proto/admin/namespace.proto:60]: what happens if you force it if the ns is not actually ready?

File: src/query/api/v1/handler/namespace/mark_ready.go (hunk 1 above)
Comment (ryanhall07) [reply src/query/api/v1/handler/namespace/mark_ready.go:51]: don't actually have a strong opinion about this. just wanted to know your thoughts on it

Comment (nbroyles) [reply src/query/api/v1/handler/namespace/mark_ready.go:51]: Ah yeah, that's a fair point. I'll update this to ready

File: src/query/generated/proto/admin/namespace.proto (hunk 3 above)
Comment (nbroyles) [reply src/query/generated/proto/admin/namespace.proto:60]: The coordinator will receive errors. I'll update the comment here to make that more explicit

File: src/query/api/v1/handler/namespace/mark_ready.go (hunk 2 above)
Comment (nbroyles) [reply src/query/api/v1/handler/namespace/mark_ready.go:56]: Yeah, actually, think that's a better idea. Will pull this out.

File: src/query/api/v1/handler/namespace/mark_ready.go (hunk 1 above)
Comment (nbroyles) [reply src/query/api/v1/handler/namespace/mark_ready.go:51]: > don't actually have a strong opinion about this. just wanted to know your thoughts on it

I wasn't in love with `mark_ready` to begin with, so this feels a bit cleaner, IMO

Comment (arnikola) [reply src/query/api/v1/handler/namespace/mark_ready.go:51]: :+1: on `ready`

File: src/query/api/v1/handler/namespace/mark_ready.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		InitialBackoff: time.Millisecond * 500,
+		MaxBackoff:     time.Second * 2,
+		BackoffFactor:  2,
+		MaxRetries:     2,
+	}
+)
+
+// MarkReadyHandler is the handler for marking namespaces ready.
+type MarkReadyHandler Handler
```
Comment (arnikola) [inline src/query/api/v1/handler/namespace/mark_ready.go:65]: Ugh this `Handler` was a bit confusing; know it's not really part of this PR, but any chance that can be renamed to `namespaceHandler`? (think we can get away without exporting it)

Comment (arnikola) [reply src/query/api/v1/handler/namespace/mark_ready.go:65]: Hm actually reading it a bit more we have annoying stuff like `nsRegistry, err = h.namespaceAddHandler.Add(namespaceRequest, opts)` which requires a base `AddHandler` type, probably fine to leave this since it's a little annoying to add interfaces to all of these?

File: src/query/generated/proto/admin/namespace.proto [hunk 5]
```diff
@@ -49,3 +49,20 @@ message NamespaceSchemaResetRequest {
 
 message NamespaceSchemaResetResponse {
 }
+
+// NamespaceReadyRequest is the request for transitioning a namespace
+// to the ready state.
+message NamespaceReadyRequest {
```
Comment (ryanhall07) [inline src/query/generated/proto/admin/namespace.proto:55]: might be nice to document this endpoint/request as idempotent so clients can safely retry any errors

**Review by ryanhall07 — APPROVED** [review]
